torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -319,46 +330,65 @@ void ProcessGroupAgent::enqueueSend(SendWork work) {
   // NB: this can be changed to use a native move capture when moved to C++14
   threadPool_.run(std::bind(
       [this](const SendWork& work) {
-        std::string serializedPayload =
-            wireSerialize(work.message_.payload(), work.message_.tensors());
-
-        std::vector<torch::Tensor> preamble = {torch::tensor(
-            {(int64_t)pg_->getRank(),
-             (int64_t)serializedPayload.length(),
-             (int64_t)work.message_.type(),
-             (int64_t)work.message_.id()},
-            {torch::kInt64})};
-
-        // ProcessGroup is not thread-safe when sending with the same tag, hence
-        // the lock
-        std::vector<std::shared_ptr<c10d::ProcessGroup::Work>> pendingSends;
-        const auto dst = work.to_.id_;
-        if (work.message_.isShutdown()) {
-          pendingSends.reserve(1);
-          {
-            std::lock_guard<std::mutex> guard(sendMutexes_[dst]);
-            pendingSends.emplace_back(
-                pg_->send(preamble, dst, dst /* channelTag */));
+        try {
+          std::string serializedPayload =
```

Comment:
Naive question: what sorts of exceptions do we expect in these codepaths?

My first expectation would be in enqueueRecv() path, where we're importing bits received from other sources, rather than on the send side. Should we handle this case first?

My other question here is whether we can really recover from some of these cases where we are catching... 

As an example, for instance, I can totally see how we could recover from a failure in deserialize()ing - we just log an error, throw away, and read the next incoming blob.

But on these order-dependent socket operations, if we're just catching and not sure what we're really recovering from, maybe it's best to just let things crash as-is, and deep dive into various point cases, rather than bandaid over a problem that we don't yet understand? (or maybe we do?)

Reply:
From the corresponding issue https://github.com/pytorch/pytorch/issues/25516, it looks like the high-level goal is to be able to satisfy the future (with an error indicating the exception) when an exception occurs in the lambdas run by the threadPool. 

One exception that we can get in `enqueueSend` (and the one that the unit test tests) is when the remote side hangs up, and `pendingSend->wait()` in `enqueueSend` throws, in this case, we can catch the exception and satisfy the future with an error indicating that the remote side has hung up.

I think for `enqueueRecv`, we should definitely handle the case you mentioned, though it also appears we should handle it in enqueueSend, rather than just swallowing the exception and resulting in the future hanging, as it may give more data to the user as to what went wrong.

Reply:
Thanks for elaborating. Agree with pendingSend->wait() throwing is a condition that we probably want to catch and handle better than we're doing now.

(That said - it's still not clear how much better we'll be able to recover at a high level, since a failure will probably cause everybody to crash now. But maybe at some point... :) )

My earlier concern was that we would try to tackle cases like where the first pendingSend goes out,  but not the other. We have basically no chance of handling those cases well - they all boil down to dropping the connection, and hoping for the best.